### PR TITLE
ClusterLoader - density test update

### PR DESCRIPTION
**Pull request by krzysied**
Adding multiple iterations of latency pods creation.

### CODE REVIEW

File: clusterloader2/pkg/framework/client/objects.go [hunk 1]
```diff
@@ -172,7 +172,9 @@ func PatchObject(dynamicClient dynamic.Interface, namespace string, name string,
 func DeleteObject(dynamicClient dynamic.Interface, gvk schema.GroupVersionKind, namespace string, name string) error {
 	gvr, _ := meta.UnsafeGuessKindToResource(gvk)
 	deleteFunc := func() error {
-		return dynamicClient.Resource(gvr).Namespace(namespace).Delete(name, nil)
+		falseVar := false
+		deleteOption := &metav1.DeleteOptions{OrphanDependents: &falseVar}
```
Comment (wojtek-t) [inline clusterloader2/pkg/framework/client/objects.go]: This means that deletion is no longer a default k8s deletion.

I'm fine with that semantic, but we need to be very clear about it:
- in comments in this file
- in design doc (it should be updated together with this PR)
- additionally in readme too

File: clusterloader2/testing/density/config.yaml [hunk 2]
```diff
@@ -1,11 +1,14 @@
 #Constants
 {{$NODES_PER_NAMESPACE := 100}}
 {{$PODS_PER_NODE := 30}}
+{{$MIN_LATENCY_PODS := 500}}
```
Comment (wojtek-t) [inline clusterloader2/testing/density/config.yaml]: Why do we make it constant?

File: clusterloader2/testing/density/config.yaml [hunk 3]
```diff
@@ -69,12 +72,39 @@ steps:
     Params:
       action: start
       labelSelector: kind = latencyPod
+{{range $iteration := (RangeInt 1 $latencyPodsIterations)}}
```
Comment (wojtek-t) [inline clusterloader2/testing/density/config.yaml]: I hate it - it starts looking like coding in yaml.

Why do we need iterations?
Why can't we simply created all latency pods in a single iteration?
I don't see any reason why we would need iterations.

Comment (krzysied) [reply clusterloader2/testing/density/config.yaml]: > I hate it - it starts looking like coding in yaml.

I agree. However I'm not sure what's the alternative.

> Why do we need iterations?

I would like to implement (more or less) the same logic as in original test - this way it's easier to compare the results.

> Why can't we simply created all latency pods in a single iteration?

That's tricky. Assuming that we already have a lot of saturation pods, adding 1 pod is not a problem. Adding 5 can exceed the quota. (Needs to be verified.)
It also depends on what we want to achieve and what we want to measure. Creating 5 times 20 pods and deleting them is different than creating and deleting 100 pods.

Comment (wojtek-t) [reply clusterloader2/testing/density/config.yaml]: So I can tell you what the motivation for that was.
In general, if there were two few pods, a single pods may affect 99th percentile. So we want to schedule bit enough number of those to avoid such situation.

TBH, I'm not convinced that what exactly was done is the best way. And while in general I agree replicating the test existing is better, we should do it carefully.

In this particular case, I really think that we don't want to do iteration - this precedent is too dangerous.
If you remember, I already told you in the past that I'm against such configs, and I would prefer hardcoding stuff. But I gave up.
But this is waaay to much for me.

re quota issues - it's about maxPodsPerNode, not quota.
And for all tests we run, we would be fine even with 5 pods.

Let's just change it to schedule max(500, #nodes) latency pods and not introduce iterations.

In fact, I still believe this test is too complicated due to all those fancy parameters that noone will never change...

File: clusterloader2/pkg/framework/client/objects.go (hunk 1 above)
Comment (krzysied) [reply clusterloader2/pkg/framework/client/objects.go]: Added in 3rd commit.

File: clusterloader2/testing/density/config.yaml (hunk 3 above)
Comment (krzysied) [reply clusterloader2/testing/density/config.yaml]: Iteration removed.
The only problem is that max(500, #nodes) works really bad if someone uses it with 5-nodes cluster.

Comment (wojtek-t) [reply clusterloader2/testing/density/config.yaml]: This test has other problems for small clusters...

Let's add a comment that this test is supposed to run on clusters larger than X (50?, 100?) and that's fine.

Comment (krzysied) [reply clusterloader2/testing/density/config.yaml]: Comment added.

> This test has other problems for small clusters...

The results was meaningless in terms of scalability - the sample was to small, right?
However, it was good as an example - everyone could run it on any cluster.

File: clusterloader2/testing/density/config.yaml [hunk 4]
```diff
@@ -1,11 +1,17 @@
...
+
 #Constants
 {{$NODES_PER_NAMESPACE := 100}}
 {{$PODS_PER_NODE := 30}}
+{{$MIN_LATENCY_PODS := 500}}
 #Variables
 {{$totalPods := MultiplyInt $PODS_PER_NODE .Nodes}}
 {{$namespaces := DivideInt (SubtractInt .Nodes 1) $NODES_PER_NAMESPACE | AddInt 1}}
 {{$replicas := DivideInt $totalPods $namespaces}}
 {{$totalPodsRounded := MultiplyInt $namespaces $replicas}}
```
[the remark this answers is not in the document]
Comment (krzysied) [reply clusterloader2/testing/density/config.yaml]: > The first thing that I would do is to understand why we really want all those Timers?

The main reasons are:
- it creates json file with all timers measurements. It can be easily parsed.
- allows to measure execution time between any two points.
- allows to connect execution time with a label. Even if we add timers in background to every phase, we still want to name these phases.

> One quick observation is that if you want to measure all phases, maybe that should just happen in the background.

I would be OK with this solution. However it's only because I know how cluster loader works.
IMHO if we want cluster loader to user friendly, we should have more explicit declaration approach instead of things running in the background.

[the remark this answers is not in the document]
Comment (wojtek-t) [reply clusterloader2/testing/density/config.yaml]: What I'm saying is that I don't see real benefits from having those (at least not enough benefits to justify 25% bigger config).
1. We are not tracking those times anywhere now
2. Just having a log" Phase %s started", "phase %s ended" is more than we need now.

I agree that if we want to track latency of some phase, it's useful. But we really don't care about it now.

File: clusterloader2/pkg/framework/client/objects.go [hunk 5]
```diff
@@ -172,7 +172,10 @@ func PatchObject(dynamicClient dynamic.Interface, namespace string, name string,
 func DeleteObject(dynamicClient dynamic.Interface, gvk schema.GroupVersionKind, namespace string, name string) error {
 	gvr, _ := meta.UnsafeGuessKindToResource(gvk)
 	deleteFunc := func() error {
-		return dynamicClient.Resource(gvr).Namespace(namespace).Delete(name, nil)
+		// Delete operation removes object with all of the dependants.
+		falseVar := false
+		deleteOption := &metav1.DeleteOptions{OrphanDependents: &falseVar}
+		return dynamicClient.Resource(gvr).Namespace(namespace).Delete(name, deleteOption)
```
Comment (wojtek-t) [inline clusterloader2/pkg/framework/client/objects.go]: Can you split all of the above to a separate PR?

I think this is much less controversial.
Wrt to changes to the test itself, I think it requires significantly more work.

Comment (krzysied) [reply clusterloader2/pkg/framework/client/objects.go]: Code changes moved to #255.

**Comment by krzysied** [comment]
/hold
Waiting for #255 to be merged.

File: clusterloader2/testing/density/config.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (krzysied) [reply clusterloader2/testing/density/config.yaml]: > We are not tracking those times anywhere now

There is a problem with Timer name - I need to revert it's name to `TestPhaseTimer` (backward compatibility). If I do this, the results will appear on perfdash.

> Just having a log" Phase %s started", "phase %s ended" is more than we need now

Just to clarify, what is an example of a text that %s will be replaced with?

[the remark this answers is not in the document]
Comment (wojtek-t) [reply clusterloader2/testing/density/config.yaml]: Discussed offline.
To summarize:
- let's introduce an optional "name" field to Step
- if it is defined, fire a timer for that step automatically

File: clusterloader2/testing/density/config.yaml [hunk 6]
```diff
@@ -90,7 +96,27 @@ steps:
...
+# Remove latency pods
+- phases:
+  - namespaceRange:
+      min: 1
+      max: {{$namespaces}}
+    replicasPerNamespace: 0
+    tuningSet: Uniform5qps
+    objectBundle:
+    - basename: latency-pod-rc
+      objectTemplatePath: rc.yaml
```
Comment (wojtek-t) [inline clusterloader2/testing/density/config.yaml]: If replicasPerNamespace is 0 (or in general smaller than before so that it results in deletion of objects),
do we really need to pass objectTemplatePath and templateFillMap?
Conceptually this shouldn't be needed.

Comment (krzysied) [reply clusterloader2/testing/density/config.yaml]: > do we really need to pass objectTemplatePath (...)?

Yes. Previously there was `ObjectKind` structure that described kind and api group. Currently this data has to be taken from yaml file. It required due to possible names collision.

> do we really need to pass (...) templateFillMap?

No. It's not required. I've removed it from test.

Comment (wojtek-t) [reply clusterloader2/testing/density/config.yaml]: OK thanks. That makes sense.

File: clusterloader2/testing/density/config.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (krzysied) [reply clusterloader2/testing/density/config.yaml]: Adding "name" field - #257

[the remark this answers is not in the document]
Comment (krzysied) [reply clusterloader2/testing/density/config.yaml]: Adding timer - #258

[the remark this answers is not in the document]
Comment (wojtek-t) [reply clusterloader2/testing/density/config.yaml]: Well - it's not timer per se, just some logs. But yes - it's definitely enough (at least for now).

File: clusterloader2/testing/density/config.yaml [hunk 7]
```diff
@@ -1,11 +1,17 @@
+# This test should be run on cluster with 100+ nodes!
+
 #Constants
 {{$NODES_PER_NAMESPACE := 100}}
 {{$PODS_PER_NODE := 30}}
+{{$MIN_LATENCY_PODS := 500}}
 #Variables
 {{$totalPods := MultiplyInt $PODS_PER_NODE .Nodes}}
 {{$namespaces := DivideInt (SubtractInt .Nodes 1) $NODES_PER_NAMESPACE | AddInt 1}}
```
[the remark this answers is not in the document]
Comment (krzysied) [reply clusterloader2/testing/density/config.yaml]: Assumption added.
It solved the problem with saturation pods, but latency pods still require rounding (due to the `MaxInt` being used).

**Comment by krzysied** [comment]
@wojtek-t 
Code is rebased. All changes from other PRs are applied.

File: clusterloader2/testing/density/config.yaml [hunk 8]
```diff
@@ -1,11 +1,15 @@
+# This test should be run on cluster with 100+ nodes!
```
Comment (wojtek-t) [inline clusterloader2/testing/density/config.yaml]:
```
# ASSUMPTIONS:
- Underlying cluster should have 100+ nodes.
- Number of nodes should be divisible by NODES_PER_NAMESPACE (default 100).
```

File: clusterloader2/testing/density/config.yaml [hunk 9]
```diff
@@ -37,16 +41,16 @@ steps:
...
       templateFillMap:
-        Replicas: {{$replicas}}
+        Replicas: {{$PODS_PER_NAMESPACE}}
         PodKind: saturationPod
 - measurements:
   - Identifier: WaitForSaturationRunning
     Method: WaitForRunningPods
     Params:
-      desiredPodCount: {{$totalPodsRounded}}
+      desiredPodCount: {{$totalPods}}
```
Comment (wojtek-t) [inline clusterloader2/testing/density/config.yaml]: I'm not convinced we should do this - I think that more standard our tests are as a whole, the better.

As a result, I would prefer using the same measurements in both tests when possible.
So I think we should use:
WaitForControlledPodsRunning (given that we have RCs here anyway).

And then you won't need totalPods'
The same observation is for latencyPods.

Comment (krzysied) [reply clusterloader2/testing/density/config.yaml]: Ok. I see your point.
I assume that this PR is blocked until #264 is merged. Or should I use the old version of WaitForControlledPodsRunning?

Comment (wojtek-t) [reply clusterloader2/testing/density/config.yaml]: Yes - it's blocked.

**Comment by krzysied** [comment]
#264 merged. PR rebased.

**Comment by krzysied** [comment]
I applied changes from #264 . Currently we are waiting for RCs being ready instead of waiting just for pods. There are two separate measurement instances. One for saturation RCs, the other for latency RCs.  I also changed the operationTimeout for saturation RCs, due to 20pods/s assumption.
@wojtek-t PTAL

File: clusterloader2/testing/density/config.yaml (hunk 8 above)
Comment (wojtek-t) [reply clusterloader2/testing/density/config.yaml]: please fix

File: clusterloader2/testing/density/config.yaml [hunk 10]
```diff
@@ -55,42 +67,60 @@ steps:
...
+- measurements:
+  - Identifier: PodStartupLatency
+    Method: PodStartupLatency
+    Params:
+      action: gather
+- measurements:
+  - Identifier: WaitForRunningLatencyRCs
+    Method: WaitForControlledPodsRunning
+    Params:
+      action: stop
```
Comment (wojtek-t) [inline clusterloader2/testing/density/config.yaml]: Do we need to stop measurements explicitly?
This is additional 5 lines of the code (2 times) that are boilerplate.

If we need, can we introduce an action "gather-and-stop" (or sth like that)?

File: clusterloader2/testing/density/rc.yaml [hunk 11]
```diff
@@ -10,7 +12,9 @@ spec:
     metadata:
       labels:
         name: {{.Name}}
+{{if .PodKind}}
```
Comment (wojtek-t) [inline clusterloader2/testing/density/rc.yaml]: No ifs please. Seriously - this is making this unreadable.

Comment (wojtek-t) [reply clusterloader2/testing/density/rc.yaml]: We should reuse the same label. Maybe sth like:
group: {{.Group}}

[where group is saturation or latency]?

File: clusterloader2/testing/density/config.yaml (hunk 10 above)
Comment (krzysied) [reply clusterloader2/testing/density/config.yaml]: I think in this case should be ok to not stop the measurement, although I'm not a fan of this approach.

I would prefer to add `Finalize()` function to measurement interface. This function would be called for every measurement instance during cleanup phase. Wdyt?

File: clusterloader2/testing/density/config.yaml (hunk 8 above)
Comment (krzysied) [reply clusterloader2/testing/density/config.yaml]: Done.

File: clusterloader2/testing/density/config.yaml (hunk 10 above)
Comment (wojtek-t) [reply clusterloader2/testing/density/config.yaml]: > I would prefer to add Finalize() function to measurement interface. This function would be called for every measurement instance during cleanup phase. Wd

This also sounds fine.

File: clusterloader2/testing/density/rc.yaml (hunk 11 above)
Comment (krzysied) [reply clusterloader2/testing/density/rc.yaml]: `if` removed.

File: clusterloader2/testing/density/config.yaml [hunk 12]
```diff
@@ -1,11 +1,19 @@
...
 #Variables
-{{$totalPods := MultiplyInt $PODS_PER_NODE .Nodes}}
-{{$namespaces := DivideInt (SubtractInt .Nodes 1) $NODES_PER_NAMESPACE | AddInt 1}}
-{{$replicas := DivideInt $totalPods $namespaces}}
-{{$totalPodsRounded := MultiplyInt $namespaces $replicas}}
+{{$namespaces := DivideInt .Nodes $NODES_PER_NAMESPACE}}
+{{$totalPods := MultiplyInt $PODS_PER_NAMESPACE $namespaces}}
+{{$latencyReplicas := DivideInt (MaxInt $MIN_LATENCY_PODS .Nodes) $namespaces}}
+{{$totalLatencyPods := MultiplyInt $namespaces $latencyReplicas}}
+{{$saturationRCTimeout := MaxInt $MIN_SATURATION_PODS_TIMEOUT_S (DivideFloat $totalPods $POD_THROUGHPUT_PER_S)}}
```
Comment (wojtek-t) [inline clusterloader2/testing/density/config.yaml]: That is too short:
https://github.com/kubernetes/kubernetes/blob/master/test/e2e/scalability/density.go#L637

File: clusterloader2/testing/density/config.yaml [hunk 13]
```diff
@@ -1,11 +1,19 @@
+# ASSUMPTIONS:
+# - Underlying cluster should have 100+ nodes.
+# - Number of nodes should be divisible by NODES_PER_NAMESPACE (default 100).
+
 #Constants
 {{$NODES_PER_NAMESPACE := 100}}
-{{$PODS_PER_NODE := 30}}
+{{$PODS_PER_NAMESPACE := 3000}}
+{{$MIN_LATENCY_PODS := 500}}
+{{$MIN_SATURATION_PODS_TIMEOUT_S := 120}}
```
Comment (wojtek-t) [inline clusterloader2/testing/density/config.yaml]: I don't think that's needed.

File: clusterloader2/testing/density/config.yaml [hunk 14]
```diff
@@ -1,11 +1,19 @@
...
+{{$POD_THROUGHPUT_PER_S := 20}}
```
Comment (wojtek-t) [inline clusterloader2/testing/density/config.yaml]: TEST_THROUGHPUT ?

File: clusterloader2/testing/density/config.yaml (hunk 13 above)
Comment (krzysied) [reply clusterloader2/testing/density/config.yaml]: I'm still using it to calculate saturationRCTimeout. Value changed to 180.

File: clusterloader2/testing/density/config.yaml (hunk 14 above)
Comment (krzysied) [reply clusterloader2/testing/density/config.yaml]: Done.

File: clusterloader2/testing/density/config.yaml [hunk 15]
```diff
@@ -1,11 +1,19 @@
+# ASSUMPTIONS:
+# - Underlying cluster should have 100+ nodes.
+# - Number of nodes should be divisible by NODES_PER_NAMESPACE (default 100).
+
 #Constants
 {{$NODES_PER_NAMESPACE := 100}}
-{{$PODS_PER_NODE := 30}}
+{{$PODS_PER_NAMESPACE := 3000}}
+{{$MIN_LATENCY_PODS := 500}}
+{{$MIN_SATURATION_PODS_TIMEOUT := 180}}
```
Comment (wojtek-t) [inline clusterloader2/testing/density/config.yaml]: I wouldn't call that min. because we add it every time.
OTOH, I don't have suggestion for better name...

File: clusterloader2/testing/density/config.yaml [hunk 16]
```diff
@@ -37,16 +53,14 @@ steps:
     - basename: saturation-rc
       objectTemplatePath: rc.yaml
       templateFillMap:
-        Replicas: {{$replicas}}
+        Replicas: {{$PODS_PER_NAMESPACE}}
+        RCKind: saturationRC
```
Comment (wojtek-t) [inline clusterloader2/testing/density/config.yaml]: I think I wasn't clear enough. I don't want to have two separate labels - this should be the same label.
It's very common pattern that you propagate a bunch of labels of grouping object to the pod template.

We should reuse the same label (let's call it "group") and set it to "saturation" or "latency".

File: clusterloader2/testing/density/config.yaml [hunk 17]
```diff
@@ -55,42 +69,55 @@ steps:
       labelSelector: kind = latencyPod
 - name: Creating saturation pods
 # Create latency pods
+- measurements:
```
Comment (wojtek-t) [inline clusterloader2/testing/density/config.yaml]: For consistency - the comment should be before the measurement.

File: clusterloader2/testing/density/config.yaml [hunk 18]
```diff
@@ -37,16 +53,14 @@ steps:
     - basename: saturation-rc
```
Comment (wojtek-t) [inline clusterloader2/testing/density/config.yaml]: I don't see a good reason for introducing "Uniform100qps" tuningset.
To simplify the config, let's remove this and use Uniform5qps here too.

File: clusterloader2/testing/density/config.yaml [hunk 19]
```diff
@@ -1,11 +1,19 @@
...
-{{$totalPods := MultiplyInt $PODS_PER_NODE .Nodes}}
-{{$namespaces := DivideInt (SubtractInt .Nodes 1) $NODES_PER_NAMESPACE | AddInt 1}}
-{{$replicas := DivideInt $totalPods $namespaces}}
-{{$totalPodsRounded := MultiplyInt $namespaces $replicas}}
+{{$namespaces := DivideInt .Nodes $NODES_PER_NAMESPACE}}
+{{$totalPods := MultiplyInt $PODS_PER_NAMESPACE $namespaces}}
+{{$latencyReplicas := DivideInt (MaxInt $MIN_LATENCY_PODS .Nodes) $namespaces}}
+{{$totalLatencyPods := MultiplyInt $namespaces $latencyReplicas}}
+{{$saturationRCTimeout := DivideFloat $totalPods $TEST_THROUGHPUT | AddInt $MIN_SATURATION_PODS_TIMEOUT}}
 #Test
```
Comment (wojtek-t) [inline clusterloader2/testing/density/config.yaml]: s/#Test/empty line/
[or maybe two empty lines?]

File: clusterloader2/testing/density/config.yaml (hunk 17 above)
Comment (krzysied) [reply clusterloader2/testing/density/config.yaml]: Done.

File: clusterloader2/testing/density/config.yaml (hunk 18 above)
Comment (krzysied) [reply clusterloader2/testing/density/config.yaml]: OK. Done.

File: clusterloader2/testing/density/config.yaml (hunk 19 above)
Comment (krzysied) [reply clusterloader2/testing/density/config.yaml]: Done.
